File: leetcode/threeSumA.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        r = []
        s = set(nums)

        for x in s:
            if x==0:
                if nums.count(x)>3:
                    r +=[(0,0,0)]
            elif nums.count(x)>1 and (-2*x in nums):
                r += [(x,x,-2*x)]
        l = list(s)
        l.sort()
        p = [x for x in l if x>0 ]
        n = [x for x in l if x<0 ]
        if 0 in l:
            sp = set(p).intersection([-x for x in n])
            for x in sp:
                r += [(-x,0,x)]
        r += AnBp(n,p).do().result()
        r += ApBn(p,n).do().result()
        rr = [sorted(x) for x in r]
        rr.sort()
        return rr

def analysis(nums):
    return  { 'E':[x for x in nums if (x%2)==0]
            , 'O':[x for x in nums if (x%2)==1]
            }

class TwoPart:
    def __init__(self,A,B):
        self.A = A
        self.B = B
        self.R = []

    def do(self):
        if len(self.A)<1:
            return self
        if len(self.B)<2:
            return self
        if len(self.B)==2:
            for x in self.A:
                if x == sum(self.B):
                    self.R += [(x,self.B[0],self.B[1])]
            return self

        LA = analysis(self.A)
        LB = analysis(self.B)

        self.R += AEBE(LA['E'],LB['E']).do().result()
        self.R += AEBO(LA['E'],LB['O']).do().result()
        self.R += AOBECO(LA['O'],LB['E'],LB['O']).do().result()
        return self


class AnBp(TwoPart):  
    """
    " A -
    " B +
    """
    def __init__(self,A,B):
        TwoPart.__init__(self,[-x for x in A],B)
    def result(self):
        return [(-a,b,c) for (a,b,c) in self.R]

class ApBn(TwoPart):
    """
    " A +
    " B -
    """
    def __init__(self,A,B):
        TwoPart.__init__(self,A,[-x for x in B])
    def result(self):
        return [(a,-b,-c) for (a,b,c) in self.R]

class AEBO(TwoPart):
    """
    " A/2+1
    " (B+1)/2
    """
    def __init__(self,A,B):
        TwoPart.__init__(self,[int(x/2)+1 for x in A],[int((x+1)/2) for x in B])
    def result(self):
        return [((a-1)*2,2*b-1,2*c-1) for (a,b,c) in self.R]

class AEBE(TwoPart):
    """
    " A/2
    " B/2
    """
    def __init__(self,A,B):
        TwoPart.__init__(self,[int(x/2) for x in A],[int(x/2) for x in B])
    def result(self):
        return [(a*2,2*b,2*c) for (a,b,c) in self.R]

class ThreePart:
    """
    " A
    " B
    " C
    """
    def __init__(self,A,B,C):
        self.A = A
        self.B = B
        self.C = C
        self.R = []

    def do(self):
        LLL = len(self.A)*len(self.B)*len(self.C)
        if LLL < 8*8*8:
            if LLL == 0:
                return self
            for x in set(self.B):
                for y in set(self.C):
                    if x+y in self.A:
                        self.R += [(x+y,x,y)]
            return self

        LA = analysis(self.A)
        LB = analysis(self.B)
        LC = analysis(self.C)

        self.R += AEBECE(LA['E'],LB['E'],LC['E']).do().result()
        self.R += AEBOCO(LA['E'],LB['O'],LC['O']).do().result()
        self.R += AOBECO(LA['O'],LB['E'],LC['O']).do().result()
        self.R += AOBOCE(LA['O'],LB['E'],LC['O']).do().result()
        return self



class AEBECE(ThreePart):
    """
    " A/2
    " B/2
    " C/2
    """
    def __init__(self,A,B,C):
        ThreePart.__init__(self,[int(x/2) for x in A],[int(x/2) for x in B],[int(x/2) for x in C])
    def result(self):
        return [(a*2,2*b,2*c) for (a,b,c) in self.R]

class AEBOCO(ThreePart):
    """
    " A/2+1
    " (B+1)/2
    " (C+1)/2
    """
    def __init__(self,A,B,C):
        ThreePart.__init__(self,[int(x/2)+1 for x in A],[int((x+1)/2) for x in B],[int((x+1)/2) for x in C])
    def result(self):
        return [((a-1)*2,2*b-1,2*c-1) for (a,b,c) in self.R]

class AOBECO(ThreePart):
    """
    " (A+1)/2
    " B/2
    " (C+1)/2
    """
    def __init__(self,A,B,C):
        ThreePart.__init__(self,[int((x+1)/2) for x in A],[int(x/2) for x in B],[int((x+1)/2) for x in C])
    def result(self):
        return [(a*2-1,2*b,2*c-1) for (a,b,c) in self.R]

class AOBOCE(ThreePart):
    """
    " (A+1)/2
    " (B+1)/2
    " C/2
    """
    def __init__(self,A,B,C):
        ThreePart.__init__(self,[int((x+1)/2) for x in A],[int((x+1)/2) for x in B],[int(x/2) for x in C])
    def result(self):
        return [(a*2-1,2*b-1,2*c) for (a,b,c) in self.R]

#l = [3,0,-2,-1,1,2]
#l = [0,0]
import dataset
l = dataset.l2
#l = [-1,-2,-3,4,1,3,0,3,-2,1,-2,2,-1,1,-5,4,-3]
a = Solution()
r = a.threeSum(l)
print(r)
print(dataset.l2_res)
print(list(zip(r[:len(dataset.l2_res)],dataset.l2_res)))

print(len(dataset.l2_res),"  ",len(r))

File: leetcode/threeSumA.cpp
#include <set>
#include <map>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
    class TwoPart {
    private:
        vector<int>* m_A;
        vector<int>* m_B;
    public:
        vector<vector<int> > m_R;
        
        TwoPart(){};

        TwoPart & init(vector<int>& A, vector<int>& B) {
            m_A = &A;
            m_B = &B;
            return *this;
        };
        
        typedef pair<int,int> BC;
        typedef map<int,set<BC> > BAC;
        TwoPart & done() {
            if(m_A->size()<1 || m_B->size()<2)
                return *this;
            BAC bc;
            int maxA = *(m_A->end()-1);
            for( auto it = m_B->begin(); it != m_B->end()-1; ++it ) {
                for( auto jt = it+1; jt != m_B->end(); ++jt ) {
                    int s = *it + *jt;
                    if( s>maxA )
                        break;
                    auto kt = bc.find(s);
                    if( kt==bc.end() ) {
                        bc[s] = set<BC>();
                        bc[s].insert(BC(*it,*jt));
                    } else {
                        kt->second.insert(BC(*it,*jt));
                    }
                }
            }
            for( auto x : *m_A ) {
                auto kt = bc.find(x);
                if(kt != bc.end()) {
                    for( auto y : kt->second ) {
                        m_R.push_back(vector<int>{x,y.first,y.second});
                    }
                }
            }
            return *this;
        };
        
        virtual vector<vector<int> >& result() {
            return m_R;
        }; 
    };

    class AnBp : public TwoPart {
    public:
        AnBp(){};
        AnBp& init(vector<int>& A, vector<int>& B) {
            for(int& i : A) 
                i = -i;
            sort(A.begin(),A.end());
            TwoPart::init(A,B);
            return *this;
        };
        vector<vector<int> >& result() {
            for(auto& x : m_R) 
                x[0] = -x[0];
            return m_R;
        };
    };

    class ApBn : public TwoPart {
    public:
        ApBn(){};
        ApBn& init(vector<int>& A, vector<int>& B) {
            for(int& i : B) 
                i = -i;
            sort(B.begin(),B.end());
            // cout << "ApBn:" << endl;
            // Solution::_dump(A);
            // Solution::_dump(B);
            TwoPart::init(A,B);
            return *this;
        };
        vector<vector<int> >& result() {
            for(auto& x : m_R) {
                x[1] = -x[1];
                x[2] = -x[2];
            }
            return m_R;
        };
    };

    class preProcessor {
    private:
        set<int> m_key;
        set<int> contain_zero;
        set<int> dup;
        set<int> negative;
        set<int> positive;
        vector<vector<int> > m_R;
    public:
        set<int>& getkey() {return m_key;};
        set<int>& getP() {return positive;};
        set<int>& getN() {return negative;};
        
        vector<vector<int> >& result() {
            for(auto x : dup) {
                m_R.push_back(vector<int>{-2*x,x,x});
            }
            for(auto x : contain_zero) {
                m_R.push_back(vector<int>{-x,0,x});
            }
            return m_R;
        };
        preProcessor(vector<int>& nums) {
            for( auto x : nums ) 
                m_key.insert(x);
            int zeros = count(nums.begin(),nums.end(),0);
            if( zeros>2 )
                contain_zero.insert(0);
            m_key.erase(0);
            for( int x : m_key ) {
                if(x < 0) {
                    negative.insert(x);
                }
                if(x > 0) {
                    positive.insert(x);
                    if( zeros>0 )
                        if(negative.count(-x)>0)
                            contain_zero.insert(x);
                }
                if(count(nums.begin(),nums.end(),x)>1 && count(nums.begin(),nums.end(),-2*x)>0) {
                    dup.insert(x);
                }
            }
        };
        void dump() {
            cout << "m_key" << endl;
            Solution::_dump<set<int>>(m_key);
            cout << "contain_zero" << endl;
            Solution::_dump<set<int>>(contain_zero);
            cout << "dup" << endl;
            Solution::_dump<set<int>>(dup);
            cout << "negative" << endl;
            Solution::_dump<set<int>>(negative);
            cout << "positive" << endl;
            Solution::_dump<set<int>>(positive);
            Solution::dumpResult(m_R);                            
        };

    };
public:
    template <typename T>
    static void _dump(T& x) {
        for(auto y : x) {
            cout << y << " ";
        }
        cout << endl;
    };
    
    static void dumpResult(vector<vector<int> >& x) {
        for(auto y : x) {
            Solution::_dump(y);
            cout << endl;
        }
        cout << endl;
    };

    vector<int>& fconv(vector<int>& x) {
        sort(x.begin(),x.end()); 
        return x;
    };
    
    vector<vector<int> >& threeSum(vector<int>& nums) {
        AnBp anbp;
        ApBn apbn;
        vector<vector<int> >* r = new vector<vector<int> >;
        preProcessor pre(nums);
        auto n = pre.getN();
        auto p = pre.getP();
        
        for(auto x : pre.result()) {
            r->push_back(fconv(x));
        }
        // pre.dump();
        vector<int> anbpA(n.begin(), n.end());
        vector<int> anbpB(p.begin(), p.end());
        for(auto x : anbp
            .init(anbpA,anbpB)
            .done()
            .result()) 
        {
            r->push_back(fconv(x));
        }
        vector<int> apbnA(n.begin(), n.end());
        vector<int> apbnB(p.begin(), p.end());
        for(auto x : apbn
            .init(apbnB, apbnA)
            .done()
            .result()) 
        {
            r->push_back(fconv(x));
        }
        sort(r->begin(),r->end());
        return *r;    
    }
};

//vector<int> vl{0,0,0};
extern vector<int> vl;
// int l[6] = {3,0,-2,-1,1,2};
int main() {
    // vector<int> vl(&l[0],&l[0]+sizeof(l)/sizeof(int));
    // for(auto i : vl) {
    //     cout << i << endl;  
    // }
    Solution a;
    vector<vector<int> >& r = a.threeSum(vl);
    for(auto x : r) {
        cout << "[";
        for(auto y:x) {
            cout << y << ", ";
        }
        cout << "]" << endl;
    }
    cout << "total: " << r.size() << endl;
    delete &r;
}


/*
class Solution {
    class TwoPart {
    private:
        vector<int>* m_A;
        vector<int>* m_B;
    public:
        vector<vector<int> > m_R;
        
        TwoPart(){};

        TwoPart & init(vector<int>& A, vector<int>& B) {
            m_A = &A;
            m_B = &B;
            return *this;
        };
        
        typedef pair<int,int> BC;
        typedef map<int,set<BC> > BAC;
        TwoPart & done() {
            if(m_A->size()<1 || m_B->size()<2)
                return *this;
            BAC bc;
            int maxA = *(m_A->end()-1);
            for( auto it = m_B->begin(); it != m_B->end()-1; ++it ) {
                for( auto jt = it+1; jt != m_B->end(); ++jt ) {
                    int s = *it + *jt;
                    if( s>maxA )
                        break;
                    auto kt = bc.find(s);
                    if( kt==bc.end() ) {
                        bc[s] = set<BC>();
                        bc[s].insert(BC(*it,*jt));
                    } else {
                        kt->second.insert(BC(*it,*jt));
                    }
                }
            }
            for( auto x : *m_A ) {
                auto kt = bc.find(x);
                if(kt != bc.end()) {
                    for( auto y : kt->second ) {
                        m_R.push_back(vector<int>{x,y.first,y.second});
                    }
                }
            }
            return *this;
        };
        
        virtual vector<vector<int> >& result() {
            return m_R;
        }; 
    };

    class AnBp : public TwoPart {
    public:
        AnBp(){};
        AnBp& init(vector<int>& A, vector<int>& B) {
            for(int& i : A) 
                i = -i;
            sort(A.begin(),A.end());
            TwoPart::init(A,B);
            return *this;
        };
        vector<vector<int> >& result() {
            for(auto& x : m_R) 
                x[0] = -x[0];
            return m_R;
        };
    };

    class ApBn : public TwoPart {
    public:
        ApBn(){};
        ApBn& init(vector<int>& A, vector<int>& B) {
            for(int& i : B) 
                i = -i;
            sort(B.begin(),B.end());
            // cout << "ApBn:" << endl;
            // Solution::_dump(A);
            // Solution::_dump(B);
            TwoPart::init(A,B);
            return *this;
        };
        vector<vector<int> >& result() {
            for(auto& x : m_R) {
                x[1] = -x[1];
                x[2] = -x[2];
            }
            return m_R;
        };
    };

    class preProcessor {
    private:
        set<int> m_key;
        set<int> contain_zero;
        set<int> dup;
        set<int> negative;
        set<int> positive;
        vector<vector<int> > m_R;
    public:
        set<int>& getkey() {return m_key;};
        set<int>& getP() {return positive;};
        set<int>& getN() {return negative;};
        
        vector<vector<int> >& result() {
            for(auto x : dup) {
                m_R.push_back(vector<int>{-2*x,x,x});
            }
            for(auto x : contain_zero) {
                m_R.push_back(vector<int>{-x,0,x});
            }
            return m_R;
        };
        preProcessor(vector<int>& nums) {
            for( auto x : nums ) 
                m_key.insert(x);
            int zeros = count(nums.begin(),nums.end(),0);
            if( zeros>2 )
                contain_zero.insert(0);
            m_key.erase(0);
            for( int x : m_key ) {
                if(x < 0) {
                    negative.insert(x);
                }
                if(x > 0) {
                    positive.insert(x);
                    if( zeros>0 )
                        if(negative.count(-x)>0)
                            contain_zero.insert(x);
                }
                if(count(nums.begin(),nums.end(),x)>1 && count(nums.begin(),nums.end(),-2*x)>0) {
                    dup.insert(x);
                }
            }
        };
        void dump() {
            cout << "m_key" << endl;
            Solution::_dump<set<int>>(m_key);
            cout << "contain_zero" << endl;
            Solution::_dump<set<int>>(contain_zero);
            cout << "dup" << endl;
            Solution::_dump<set<int>>(dup);
            cout << "negative" << endl;
            Solution::_dump<set<int>>(negative);
            cout << "positive" << endl;
            Solution::_dump<set<int>>(positive);
            Solution::dumpResult(m_R);                            
        };

    };
public:
    template <typename T>
    static void _dump(T& x) {
        for(auto y : x) {
            cout << y << " ";
        }
        cout << endl;
    };
    
    static void dumpResult(vector<vector<int> >& x) {
        for(auto y : x) {
            Solution::_dump(y);
            cout << endl;
        }
        cout << endl;
    };

    vector<int>& fconv(vector<int>& x) {
        sort(x.begin(),x.end()); 
        return x;
    };
    
    vector<vector<int> >& threeSum(vector<int>& nums) {
        AnBp anbp;
        ApBn apbn;
        vector<vector<int> >* r = new vector<vector<int> >;
        if(nums.size()<3)
            return *r;
        preProcessor pre(nums);
        auto n = pre.getN();
        auto p = pre.getP();
        
        for(auto x : pre.result()) {
            r->push_back(fconv(x));
        }
        // pre.dump();
        vector<int> anbpA(n.begin(), n.end());
        vector<int> anbpB(p.begin(), p.end());
        for(auto x : anbp
            .init(anbpA,anbpB)
            .done()
            .result()) 
        {
            r->push_back(fconv(x));
        }
        vector<int> apbnA(n.begin(), n.end());
        vector<int> apbnB(p.begin(), p.end());
        for(auto x : apbn
            .init(apbnB, apbnA)
            .done()
            .result()) 
        {
            r->push_back(fconv(x));
        }
        sort(r->begin(),r->end());
        return *r;    
    }
};
*/